File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/array/matrix/P54_SpiralMatrix.java
package com.dsa.leetcode.labuladong.array.matrix;

import java.util.ArrayList;
import java.util.List;

/**
 * @since 2023/12/17 17:16
 */
public class P54_SpiralMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m = matrix.length;
        int n = matrix[0].length;
        List<Integer> list = new ArrayList<>(m * n);
        // 顺时针定义四个角的边界索引
        int top = 0, left = 0;
        int bottom = m - 1, right = n - 1;
        while (top <= bottom && left <= right) {
            for (int i = left; i <= right; i++) {
                list.add(matrix[top][i]);
            }
            for (int i = top + 1; i <= bottom; i++) {
                list.add(matrix[i][right]);
            }
            if (left < right && top < bottom) {
                for (int i = right - 1; i > left; i--) {
                    list.add(matrix[bottom][i]);
                }
                for (int i = bottom; i > top; i--) {
                    list.add(matrix[i][left]);
                }
            }
            top++;
            left++;
            bottom--;
            right--;
        }
        return list;
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/array/matrix/P54_SpiralMatrix.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        ans = []
        m = len(matrix)
        n = len(matrix[0])
        top, left = 0, 0
        bottom, right = m - 1, n - 1
        while top <= bottom and left <= right:
            if top == bottom:
                for i in range(left, right + 1):
                    ans.append(matrix[top][i])
            elif left == right:
                for i in range(top, bottom + 1):
                    ans.append(matrix[i][left])
            else:
                for i in range(left, right):
                    ans.append(matrix[top][i])
                for i in range(top, bottom):
                    ans.append(matrix[i][right])
                for i in range(right, left, -1):
                    ans.append(matrix[bottom][i])
                for i in range(bottom, top, -1):
                    ans.append(matrix[i][left])
            top += 1
            left += 1
            bottom -= 1
            right -= 1
        return ans


if __name__ == '__main__':
    matrix = [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]
    solu = Solution()
    print(solu.spiralOrder(matrix))